# Profitable App Profiles for the App Store and Google Play Markets

<br />

<br />


![header.jpg](header.jpg)


## Introduction


As a data analyst working for a company that **only build free apps to download and install** available on Google Play and in the App Store, our main source of revenue consists of in-app ads. 

This means that the number of users of our apps determines our revenue for any given app therefore the more users who see and engage with the ads, the better. 


<br />
 

* * *

 Collect and analyze data from each of the online stores to understand what type of apps are likely to attract more users.

# <a id='0'>Index</a>

 
### <a href='#1'>1. Exploring data </a>
 
 
- <a href='#11'>1.1 Function to explore datasets</a>


-  <a href='#12'>1.2 Data dictionary:</a>


-  <a href='#13'>1.3 Knowing the size in both data sets.</a>

### <a href='#2'>2. Data cleaning </a>


-  <a href='#21'>2.1. Detect inaccurate data... remove it.</a>


-  <a href='#22'>2.2 Detect duplicate data, and remove the duplicates.</a>


-  <a href='#23'>2.3 What differentiates duplicate applications?</a>


-  <a href='#24'>2.4 Removing Non-English Apps</a>


-  <a href='#25'>2.5 Filter non-English applications from both datasets. </a>


-  <a href='#26'>2.6 Exploring datasets and see how many rows we have left for each dataset.</a>



### <a href='#3'>3. Isolating Free Apps </a>


-  <a href='#31'>3.1 Our **ultimate goal** </a>


-  <a href='#32'>3.2 Profiles of apps successful in both markets	Table</a>



### <a href='#4'>4. Most Popular Apps by Genre on Google Play </a>

### <a href='#5'>5. Summary. </a>

* * *

The datasets are in two documents of type csv's i.e. **Comma Separated Values file** for read the content, it is necessary to load a function called `reader` of the python `csv` module.

In [1]:
from csv import reader

* * *
<a href='#0'> back to index</a>

## <a id='1'>1. Exploring datasets</a>:

We are going to explore the two sets of data, but first we are going to kwon what type of [character encoding](https://en.wikipedia.org/wiki/Character_encoding) both datasets have.

A command called [file](https://www.man7.org/linux/man-pages/man1/file.1.html), will help us to kwon the type of file regardless of its extension and avoid a error called `UnicodeDecodeError`.

In [2]:
! file -i AppleStore.csv

AppleStore.csv: application/csv; charset=utf-8


In [3]:
! file -i googleplaystore.csv

googleplaystore.csv: application/csv; charset=utf-8


Our files have the same type of charset https://en.wikipedia.org/wiki/UTF-8.


Quick idea:

- **charset:** is the set of characters you can use.
- **encoding:** is the way these characters are stored into memory.

[source](https://stackoverflow.com/questions/2281646/whats-the-difference-between-encoding-and-charset)

In [4]:
AppleStore = open('AppleStore.csv', encoding='utf8')
AppleStore = reader(AppleStore)
AppleStore = list(AppleStore)
header_apple = [x for x in AppleStore[0]]
header_apple #Columns on AppleStore

['id',
 'track_name',
 'size_bytes',
 'currency',
 'price',
 'rating_count_tot',
 'rating_count_ver',
 'user_rating',
 'user_rating_ver',
 'ver',
 'cont_rating',
 'prime_genre',
 'sup_devices.num',
 'ipadSc_urls.num',
 'lang.num',
 'vpp_lic']

In [5]:
dataset_apple = AppleStore[1:] #whole dataset apple, no headers

In [6]:
GooglePlay = open('googleplaystore.csv', encoding='utf8')
GooglePlay = reader(GooglePlay)
GooglePlay = list(GooglePlay)
header_google = [x for x in GooglePlay[0]]
header_google 

['App',
 'Category',
 'Rating',
 'Reviews',
 'Size',
 'Installs',
 'Type',
 'Price',
 'Content Rating',
 'Genres',
 'Last Updated',
 'Current Ver',
 'Android Ver']

In [7]:
dataset_google = GooglePlay[1:] #whole dataset google, no header

* * *
<a href='#0'> back to index</a>

### <a id='11'>1.1 Function to explore datasets</a>

To make easier to explore, we created a function named `explore_data()` you can repeatedly use to print rows in a readable way.

In [8]:
def explore_data(dataset, start, end, rows_and_columns=False):
        dataset_slice = dataset[start:end]    
        for row in dataset_slice:
            print(row)
            print('\n') # adds a new (empty) line after each row
        if rows_and_columns:
            print('Number of rows:', len(dataset))
            print('Number of columns:', len(dataset[0]))

In [9]:
%%html
<style>
table {float:left}
</style>

<a href='#0'> back to index</a>
* * *
### <a id='12'> 1.2 Data dictionary:</a>

We have the name of the columns and the data that form the set of both data sets:


#### AppleStore.csv

 - `header_apple`

 - `dataset_apple` 

||Name | Description|
|:--|:---|:--|
|1|"id" : |App ID|
|2|"track_name": |App Name|
|3|"size_bytes": |Size (in Bytes)|
|4|"currency": |Currency Type|
|5|"price": |Price amount|
|6|"ratingcounttot": |User Rating counts (for all version)|
|7|"ratingcountver": |User Rating counts (for current version)|
|8|"user_rating" : |Average User Rating value (for all version)|
|9|"userratingver": |Average User Rating value (for current version)|
|10|"ver" : |Latest version code|
|11|"cont_rating": |Content Rating|
|12|"prime_genre": |Primary Genre|
|13|"sup_devices.num": |Number of supporting devices|
|14|"ipadSc_urls.num": |Number of screenshots showed for display|
|15|"lang.num": |Number of supported languages|
|16|"vpp_lic": |Vpp Device Based Licensing Enabled|


#### Googleplaystore.csv

 - `header_google`

 - `dataset_google`


||Name | Description|
|:--|:---|:--|
|1 | App: |Application name|
|2 | Category: |Category the app belongs to|
|3 | Rating: |Overall user rating of the app (as when scraped)|
|4 | Reviews: | Number of user reviews for the app (as when scraped)|
|5 | Size: | Size of the app (as when scraped)|
|6 | Installs: | Number of user downloads/installs for the app (as when scraped)|
|7 | Type: | Paid or Free|
|8 | Price: | Price of the app (as when scraped)|
|9 | Content: | Rating Age group the app is targeted at - Children / Mature 21+ / Adult|
|10 |  Genres: | An app can belong to multiple genres (apart from its main category). For eg, a musical family game will belong to Music, Game, Family genres.|

<a href='#0'> back to index</a>
* * *
### <a id='13'>1.3 Knowing the size in both data sets</a>.

In [10]:
explore_data(dataset_apple, 2, 4, True)

['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']


['420009108', 'Temple Run', '65921024', 'USD', '0.0', '1724546', '3842', '4.5', '4.0', '1.6.2', '9+', 'Games', '40', '5', '1', '1']


Number of rows: 7197
Number of columns: 16


In [11]:
explore_data(dataset_google, 2, 4, True)

['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


Number of rows: 10841
Number of columns: 13


### Resume

|`dataset_apple`|`dataset_google`|
|:----|:----|
|Number of rows: 7197| Number of rows: 10841|
|Number of columns: 16|Number of columns: 13|

<a href='#0'> back to index</a>
* * *
## <a id='2'>2. Data cleaning</a>

This means that we need to:

- 1. **Detect inaccurate data, and correct or remove it**.

- 2. **Detect duplicate data, and remove the duplicates**.


- 3. **Remove non-English apps like 爱奇艺PPS -《欢乐颂2》电视剧热播**.

- 4. **Remove apps that aren't free##**.


The Google Play dataset has a dedicated [discussion section](https://www.kaggle.com/datasets/lava18/google-play-store-apps/discussion?sort=votes), and we can see that [one of the discussions](https://www.kaggle.com/datasets/lava18/google-play-store-apps/discussion/66015) describes an error for a certain row.

<a href='#0'> back to index</a>
* * *
### <a id='21'>2.1. Detect inaccurate data... remove it.</a>

The idea that the number of columns must be the same as the number of fields contained in a row means if the number of fields in a row varies the some kind of data is missing.

This function has a pair of parameters `dataset`, `header` in which we enter the dataset and the header.

the passage of each row compare the number of fields with the number of fields in header if they do not coincide then in that line some type of data is missing.

In [12]:
def field_number(dataset, header):
    index_number = 0
    for row in range(len(dataset)):
        if len(dataset[row]) != len(header):
            print(dataset[row])
            print("\n")
            print("error in index number: ",index_number)
        index_number +=1

In [13]:
field_number(dataset_apple, header_apple)

In [14]:
field_number(dataset_google, header_google)

['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']


error in index number:  10472


In [15]:
len(dataset_google[10472]) #discussion row

12

We have passed the two datasets through the function and there is a row in the `dataset_google` with a different length from the number of columns therefore we are going to remove it from our dataset.

In [16]:
del dataset_google[10472]

Checking if row has been removed properlly.

In [17]:
dataset_google[10472]

['osmino Wi-Fi: free WiFi',
 'TOOLS',
 '4.2',
 '134203',
 '4.1M',
 '10,000,000+',
 'Free',
 '0',
 'Everyone',
 'Tools',
 'August 7, 2018',
 '6.06.14',
 '4.4 and up']

#### yes ...It has been removed

<a href='#0'> back to index</a>
* * *
### <a id='22'>2.2 Detect duplicate data, and remove the duplicates.</a>

A for runs through each row extracting the name of the application.

- checking if it already exists or not.

- if not then the name is inserted in `unique_apps` else is inserted into `duplicate_apps`.

In [18]:
def detect_duplicated(dataset):
    unique_apps = []
    duplicate_apps = []
    
    for row in dataset:
        name = row[0]
        if name in unique_apps:
            duplicate_apps.append(name)
        else:
            unique_apps.append(name)
            
    if len(duplicate_apps) == 0:
        print("In this dataset there are no duplicate apps")
    else:
        times=len(duplicate_apps)
        print("There are {reapeated} repeated apps".format(reapeated=times,dataset=dataset))
        print("\n")
        print('Examples of duplicate apps:',"\n"+"\n" , duplicate_apps[:5])

In [19]:
detect_duplicated(dataset_apple)

In this dataset there are no duplicate apps


In [20]:
detect_duplicated(dataset_google)

There are 1181 repeated apps


Examples of duplicate apps: 

 ['Quick PDF Scanner + OCR FREE', 'Box', 'Google My Business', 'ZOOM Cloud Meetings', 'join.me - Simple Meetings']


Just only 5 samples of the repeated applications in **dataset_google**

* * * 
<a href='#0'> back to index</a>
### <a id='23'>2.3 What differentiates duplicate applications?</a>

If we take as an example a repeated application, in this case `Instagram`, we see that of all the fields that make up the row **there is only one that is different**.

In [21]:
for app in dataset_google:
    name = app[0]
    if name == 'Instagram':
        print(app)

['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577446', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66509917', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']


That is the number of `'Reviews'`:


Therefore this will be the criterion we use to eliminate duplicate applications.

- Create a dictionary `reviews_max`, only `name` as a key and the value as `reviews`.

- Through the dataset row by row extracting the `name` of the application in one variable and the `reviews` in another.

- if `name` variable exists in the dictionary and the value associated with this key is less than the `n_reviews` variable then we update the value in the dictionary.

- else if the name does not exist in the dictionary then we add it along with its value which will be `n_reviews`.

In [22]:
reviews_max = {} #key=name, value=review

for row in dataset_google:
    name = row[0]
    n_reviews = float(row[3])
    
    if name in reviews_max and reviews_max[name] < n_reviews: #if name in reviews_max & value in reviews_max[name]
        reviews_max[name] = n_reviews                         #is < than n_reviews the update
        
    elif name not in reviews_max:
        reviews_max[name] = n_reviews
        
print('length of the dictionary:',len(reviews_max))

length of the dictionary: 9659


This is the new length of our dataset having deleted repeated applications.

The original length of the dataset minus the number of repeated applications results in this number.

In [23]:
print('expected length', len(dataset_google) - 1181) #1181 repeated apps

expected length 9659


- this time we create two lists

- we go through the dataset row by row

- we extract the name and the number of reviews as we had done before.

- the condition works as follows:

- if the number of `reviews` is equal to the number of reviews in the previous dictionary  (`reviews_max`) and the name is not in the list called `already_added`

- then insert the whole row in the list called `android_clean` and the **name** in the list `already_added`



In [24]:
android_clean = []
already_added = []

for row in dataset_google: #original dataset
    name = row[0]
    n_reviews = float(row[3])
    
    if n_reviews == reviews_max[name] and name not in already_added:
        android_clean.append(row) #row list added to list
        already_added.append(name) #new name added list

`android_clean` is the our **new dataset**

In [25]:
explore_data(android_clean, 1, 2, True)

['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


Number of rows: 9659
Number of columns: 13


<a href='#0'> back to index</a>
* * *
### <a id='24'>2.4 Removing Non-English Apps</a>

It is no necessary to do the same with App Store data because there are no duplicates.

We use English for the applications we develop in our company, and we would like to analyze only applications aimed at an English-speaking audience. 


However, if we explore the data enough, we will find that **both datasets** have applications with names that suggest they are not aimed at an English-speaking audience.

In [26]:
print(AppleStore[813][1])
print(AppleStore[6731][1])
print('\n')
print(android_clean[4412][0])
print(android_clean[7940][0])

BATTLE BEARS -1
Beast Poker


中国語 AQリスニング
لعبة تقدر تربح DZ


**One way to do this is to remove all applications whose name contains a symbol that is not commonly used in English text**. 

English text typically includes letters of the English alphabet, numbers composed of digits from 0 to 9, punctuation marks (., !, ?, ;) and other símbols (+, *, /).

Each character we use in a string has a corresponding number associated with it. 

For example, the corresponding number of the character"a" es 97, el del carácter "A" es 65, y el del carácter "爱" es 29.233. 

 We can get the corresponding number of each character using the built-in function **ord().**

In [27]:
print(ord('a'))
print(ord('A'))
print(ord('爱'))
print(ord('5'))
print(ord('+'))

97
65
29233
53
43


The numbers corresponding to the characters that we usually use in an English text are all in the range of 0 to 127, according to the ASCII system (American Standard Code for Information Interchange). 

Based on this range of numbers, we can construct a function that detects whether a character belongs to the common English character set or not. 

- **If the number is equal to or less than 127, the character belongs to the set of common English characters. 

- **If an app's name contains a character greater than 127, it probably means that the app has a non-English name.** 

This is the basis of operation. 

Read each of the characters and compare their ASCII value with the one we have determined as limit (127).

In [28]:
def english_speak(string):
    for character in string:
        valor_character = ord(character)
        if valor_character > 127:
            return False #not English 
        else:
            return True # is English

In [29]:
english_speak('Instagram')

True

In [30]:
english_speak('爱奇艺PPS -《欢乐颂2》电视剧热播')

False

In [31]:
english_speak('Docs To Go™ Free Office Suite')#Is English!

True

In [32]:
english_speak('Instachat 😜') #Is English!

True

In [33]:
print(ord('™'))
print(ord('😜'))

8482
128540


If we are going to use the function we have created, we will lose useful data, we saw that the function could not correctly identify certain names of English applications such as **'Docs To Go™ Free Office Suite'** and **'Instachat 😜'** since many English applications will be incorrectly labeled as non-English. 


To **minimize the impact of data loss**,it is necessary to have a basic criterion that helps in the screening so **we will only delete an application if its name is longer than three characters with the corresponding numbers outside the ASCII range**. 

This means that all English apps with up to three emoji or other special characters will still be labeled as English. 

In [34]:
def english_speak(text):
    non_validchar = [] #aux list
    for character in text:
        valor_character = ord(character)
        if valor_character > 127:
            non_validchar.append(valor_character)
    if len(non_validchar) >= 3:
        return False
    else:
        return True

In [35]:
english_speak('Docs To Go™ Free Office Suite')

True

In [36]:
english_speak('爱奇艺PPS -《欢乐颂2》电视剧热播')

False

In [37]:
english_speak('Instachat 😜 😜')

True

In [38]:
english_speak('Instachat 😜 😜 😜')

False

At this point it is important to remember what the names of the datasets are, thus avoiding failures.

|Dataset names|lenght|
|:----|:----|
|`dataset_apple`|7197|
|`android_clean`|9659, 8757|

<a href='#0'> back to index</a>
* * *
### <a id='25'>2.5 Filter non-English applications from both datasets.</a> 

If an app's name is identified as English, the we add the entire row to a separate list.

In [39]:
# AppleStore

list_apple_clean = []

for row in dataset_apple:
    name = row[1]
    if english_speak(name):
        list_apple_clean.append(row)
    else:
        pass

In [40]:
## android_clean

list_android_clean = []

for row in android_clean:
    nombre = row[0]
    if english_speak(nombre):
        list_android_clean.append(row)
    else:
        pass

<a href='#0'> back to index</a>
* * *
### <a id='26'>2.6 Exploring datasets and see how many rows we have left for each dataset.</a>

In each cleaning process we have put different names to our dataset variables so it is important to take them into account

In [41]:
explore_data(list_apple_clean, 2, 4, True)

['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']


['420009108', 'Temple Run', '65921024', 'USD', '0.0', '1724546', '3842', '4.5', '4.0', '1.6.2', '9+', 'Games', '40', '5', '1', '1']


Number of rows: 6155
Number of columns: 16


In [42]:
explore_data(list_android_clean, 2, 4, True)

['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


['Pixel Draw - Number Art Coloring Book', 'ART_AND_DESIGN', '4.3', '967', '2.8M', '100,000+', 'Free', '0', 'Everyone', 'Art & Design;Creativity', 'June 20, 2018', '1.1', '4.4 and up']


Number of rows: 9597
Number of columns: 13


**Now we are working with this:**


|Dataset names|lenght|
|:----|:----|
|`list_apple_clean`|6155|
|`list_android_clean`|9597, 8698|

<a href='#0'> back to index</a>
* * *
## <a id='3'>3. Isolating Free Apps</a>

Our datasets contain **free and non-free applications** we need to isolate only the free applications for our analysis.

Checking the length of each dataset to see how many apps have left.

In [43]:
header_apple

['id',
 'track_name',
 'size_bytes',
 'currency',
 'price',
 'rating_count_tot',
 'rating_count_ver',
 'user_rating',
 'user_rating_ver',
 'ver',
 'cont_rating',
 'prime_genre',
 'sup_devices.num',
 'ipadSc_urls.num',
 'lang.num',
 'vpp_lic']

In [44]:
free_apple_clean = []

for row in list_apple_clean:
    prix = row[4] #'price'
 
    if prix == '0.0' or prix == '0': #not numbers just characters! 
        free_apple_clean.append(row)
    else:
        pass
    
len(free_apple_clean)

3203

In [45]:
header_google

['App',
 'Category',
 'Rating',
 'Reviews',
 'Size',
 'Installs',
 'Type',
 'Price',
 'Content Rating',
 'Genres',
 'Last Updated',
 'Current Ver',
 'Android Ver']

In [46]:
free_android_clean = [] 

for row in list_android_clean:
    prix = row[7] #'Price'
    if prix == '0.0' or prix == '0': #not numbers just characters! 
        free_android_clean.append(row)
    else:
        pass
    
len(free_android_clean)

8848

**Updating names and lengths**


|Dataset names|lenght|
|:----|:----|
|`free_apple_clean`|3203|
|`free_android_clean`|8848|

<a href='#0'> back to index</a>
* * *
### <a id='31'>Our **ultimate goal** is to add the app on both Google Play and the App Store, so we need to **find profiles of apps that are successful in both markets**.</a> 


Let's start the analysis by having an idea of what are the most common genres for each market. To do this, we will have to build frequency tables for some columns of our datasets.

The columns that can give us the information we need are:

`free_apple_clean[11]` ---> `prime_genre`

`free_android_clean[]` ---> `Genres`

- create a function

- Create a dictionary `table`

- Running dataset row by row

- A counter will tell us the total number of rows

- The value that we want to extract from the column that interests us we introduce it in `value`

- if `value` exists in the dictionary `table` we increment the value in 1 but if not exist we give it a value of 1.

- We created another dictionary called `table_percentages`

- Running through the previous dictionary called `table`

- The percentage will be the result of the value from the dictionary divided by the `total` value.

- In the dictionary we enter the **key** as column and a **value** as the percentage

- return the diccionary

In [47]:
def freq_table(dataset, column):
    table = {}
    total = 0
    
    for row in dataset:
        total +=1 # counter give us length.
        value = row[column]
        if value in table:
            table[value] +=1
        else:
            table[value] = 1
        
    table_percentages = {}
    
    for key in table: #table[column value]<--as key] value <--as +1 values added 
        percentage = (table[key] / total ) * 100
        table_percentages[key]=percentage
            
    return table_percentages # {column:times repeated}

- Create a function called `display_table` same parameters as `freq_table` because this function is inside of `display_table`

- Create a list `table_display`

- Running through the `table` dictionary 

- The value of the dictionary we place in `key_val_as_tuple`, tuple along with the key of this

- Introducing tuple it in the list `table_display`

- when append the list is finish we can order it thanks to the **sorted** function

- Going through that variable with a new for we can print the result in an orderly manner

In [48]:
def display_table(dataset, column):
    table = freq_table(dataset, column) #return diccionary
    table_display = []
    for key in table:
        key_val_as_tuple = (table[key], key)   #times repeated & key same variable
        table_display.append(key_val_as_tuple) #append tuple into list

    table_sorted = sorted(table_display, reverse = True) #sorting list
    for entry in table_sorted: #print every cell on list
        print(entry[1], ':', round(entry[0],2),'%')

In [49]:
output_apple = display_table(free_apple_clean,11) #apple prime_genere column

Games : 58.26 %
Entertainment : 7.84 %
Photo & Video : 5.0 %
Education : 3.68 %
Social Networking : 3.31 %
Shopping : 2.59 %
Utilities : 2.47 %
Sports : 2.15 %
Music : 2.06 %
Health & Fitness : 2.03 %
Productivity : 1.75 %
Lifestyle : 1.56 %
News : 1.34 %
Travel : 1.25 %
Finance : 1.09 %
Weather : 0.87 %
Food & Drink : 0.81 %
Reference : 0.53 %
Business : 0.53 %
Book : 0.37 %
Navigation : 0.19 %
Medical : 0.19 %
Catalogs : 0.12 %


In [50]:
output_gogle = display_table(free_android_clean,9) #Goolge Genres column

Tools : 8.44 %
Entertainment : 6.08 %
Education : 5.36 %
Business : 4.6 %
Productivity : 3.9 %
Lifestyle : 3.88 %
Finance : 3.71 %
Medical : 3.54 %
Sports : 3.46 %
Personalization : 3.32 %
Communication : 3.23 %
Action : 3.1 %
Health & Fitness : 3.09 %
Photography : 2.95 %
News & Magazines : 2.8 %
Social : 2.67 %
Travel & Local : 2.33 %
Shopping : 2.25 %
Books & Reference : 2.14 %
Simulation : 2.05 %
Dating : 1.86 %
Arcade : 1.84 %
Video Players & Editors : 1.77 %
Casual : 1.76 %
Maps & Navigation : 1.39 %
Food & Drink : 1.24 %
Puzzle : 1.13 %
Racing : 0.99 %
Role Playing : 0.94 %
Libraries & Demo : 0.94 %
Auto & Vehicles : 0.93 %
Strategy : 0.92 %
House & Home : 0.8 %
Weather : 0.79 %
Events : 0.71 %
Adventure : 0.67 %
Comics : 0.6 %
Beauty : 0.6 %
Art & Design : 0.6 %
Parenting : 0.5 %
Card : 0.45 %
Trivia : 0.42 %
Casino : 0.42 %
Educational;Education : 0.4 %
Board : 0.38 %
Educational : 0.37 %
Education;Education : 0.34 %
Word : 0.26 %
Casual;Pretend Play : 0.24 %
Music : 0.2 %
Rac

<a href='#0'> back to index</a>
* * *
### <a id='32'>3.2 Profiles of apps successful in both markets</a>

|Dataset name|Genres|percentage|
|:---|:---|:---|
|`free_apple_clean`|||
||Games : |58 %|
||**Entertainment :** |7,84 %|
||Photo & Video : |5 %|
||Education : |4 %|
||Social Networking : |3 %|
|`free_android_clean`|||
||Tools : 8 %
||**Entertainment :**| 6.08 %|
||Education : |5 %|
||Business : |5 %|
||Productivity : |4 %|

#### We would like to get an idea about the type of apps with more users.

One way to find out which genres are the most popular (have the most users) is to calculate the **average number of installs for each app genre**.

- For the **Google Play dataset, we can find this information in the installations** column, 

- For the App Store dataset **we will take the total number of user ratings as a proxy, which we can find in the 'rating_count_column' app.**

In [51]:
genres_ios = freq_table(free_apple_clean, -5)
list_order = {}

for genre in genres_ios:
    total = 0
    len_genre = 0
    for app in free_apple_clean:
        genre_app = app[-5]
        if genre_app == genre:            
            n_ratings = float(app[5])
            total += n_ratings
            len_genre += 1
            
    avg_n_ratings = total / len_genre
    print(genre, ':', avg_n_ratings)

Social Networking : 71548.34905660378
Photo & Video : 28441.54375
Games : 22886.36709539121
Music : 57326.530303030304
Reference : 79350.4705882353
Health & Fitness : 23298.015384615384
Weather : 52279.892857142855
Utilities : 19156.493670886077
Travel : 28243.8
Shopping : 27230.734939759037
News : 21248.023255813954
Navigation : 86090.33333333333
Lifestyle : 16815.48
Entertainment : 14195.358565737051
Food & Drink : 33333.92307692308
Sports : 23008.898550724636
Book : 46384.916666666664
Finance : 32367.02857142857
Education : 7003.983050847458
Productivity : 21028.410714285714
Business : 7491.117647058823
Catalogs : 4004.0
Medical : 612.0


In [52]:
genres_ios = freq_table(free_apple_clean, -5)
list_order = {}

for genre in genres_ios:
    total = 0
    len_genre = 0
    for app in free_apple_clean:
        genre_app = app[-5]
        if genre_app == genre:            
            n_ratings = float(app[5])
            total += n_ratings
            len_genre += 1
            
    avg_n_ratings = total / len_genre
    print(genre, ':', avg_n_ratings)
    
    list_order[genre]=avg_n_ratings

Social Networking : 71548.34905660378
Photo & Video : 28441.54375
Games : 22886.36709539121
Music : 57326.530303030304
Reference : 79350.4705882353
Health & Fitness : 23298.015384615384
Weather : 52279.892857142855
Utilities : 19156.493670886077
Travel : 28243.8
Shopping : 27230.734939759037
News : 21248.023255813954
Navigation : 86090.33333333333
Lifestyle : 16815.48
Entertainment : 14195.358565737051
Food & Drink : 33333.92307692308
Sports : 23008.898550724636
Book : 46384.916666666664
Finance : 32367.02857142857
Education : 7003.983050847458
Productivity : 21028.410714285714
Business : 7491.117647058823
Catalogs : 4004.0
Medical : 612.0


In [53]:
marklist = sorted(list_order.items(), key=lambda x:x[1],reverse = True)
sortdict = dict(marklist)
sortdict

{'Navigation': 86090.33333333333,
 'Reference': 79350.4705882353,
 'Social Networking': 71548.34905660378,
 'Music': 57326.530303030304,
 'Weather': 52279.892857142855,
 'Book': 46384.916666666664,
 'Food & Drink': 33333.92307692308,
 'Finance': 32367.02857142857,
 'Photo & Video': 28441.54375,
 'Travel': 28243.8,
 'Shopping': 27230.734939759037,
 'Health & Fitness': 23298.015384615384,
 'Sports': 23008.898550724636,
 'Games': 22886.36709539121,
 'News': 21248.023255813954,
 'Productivity': 21028.410714285714,
 'Utilities': 19156.493670886077,
 'Lifestyle': 16815.48,
 'Entertainment': 14195.358565737051,
 'Business': 7491.117647058823,
 'Education': 7003.983050847458,
 'Catalogs': 4004.0,
 'Medical': 612.0}

In [54]:
max(list_order, key=list_order.get) # navigation apps have the highest number of user reviews,

'Navigation'

In [55]:
list_order['Navigation']

86090.33333333333

In [56]:
for app in free_apple_clean:
    if app[-5] == 'Navigation':
        print(app[1], ':', app[5]) # print name and number of ratings

Waze - GPS Navigation, Maps & Real-time Traffic : 345046
Google Maps - Navigation & Transit : 154911
Geocaching® : 12811
CoPilot GPS – Car Navigation & Offline Maps : 3582
ImmobilienScout24: Real Estate Search in Germany : 187
Railway Route Search : 5


The same pattern applies to social networking apps, where the average number is heavily influenced by a few giants like Facebook, Pinterest, Skype, etc. Same applies to music apps, where a few big players like Pandora, Spotify, and Shazam heavily influence the average number.

Our aim is to find popular genres, but navigation, social networking or music apps might seem more popular than they really are. The average number of ratings seem to be skewed by very few apps which have hundreds of thousands of user ratings, while the other apps may struggle to get past the 10,000 threshold. We could get a better picture by removing these extremely popular apps for each genre and then rework the averages, but we'll leave this level of detail for later.

Reference apps have 74,942 user ratings on average, but it's actually the Bible and Dictionary.com which skew up the average rating:

In [57]:
for app in free_apple_clean:
    if app[-5] == 'Reference':
        print(app[1], ':', app[5])

Bible : 985920
Dictionary.com Dictionary & Thesaurus : 200047
Dictionary.com Dictionary & Thesaurus for iPad : 54175
Google Translate : 26786
Muslim Pro: Ramadan 2017 Prayer Times, Azan, Quran : 18418
New Furniture Mods - Pocket Wiki & Game Tools for Minecraft PC Edition : 17588
Merriam-Webster Dictionary : 16849
Night Sky : 12122
City Maps for Minecraft PE - The Best Maps for Minecraft Pocket Edition (MCPE) : 8535
LUCKY BLOCK MOD ™ for Minecraft PC Edition - The Best Pocket Wiki & Mods Installer Tools : 4693
GUNS MODS for Minecraft PC Edition - Mods Tools : 1497
Guides for Pokémon GO - Pokemon GO News and Cheats : 826
WWDC : 762
Horror Maps for Minecraft PE - Download The Scariest Maps for Minecraft Pocket Edition (MCPE) Free : 718
VPN Express : 14
Real Bike Traffic Rider Virtual Reality Glasses : 8
Jishokun-Japanese English Dictionary & Translator : 0


However, this niche seems to show some potential. One thing we could do is take another popular book and turn it into an app where we could add different features besides the raw version of the book. This might include daily quotes from the book, an audio version of the book, quizzes about the book, etc. On top of that, we could also embed a dictionary within the app, so users don't need to exit our app to look up words in an external app.

This idea seems to fit well with the fact that the App Store is dominated by for-fun apps. This suggests the market might be a bit saturated with for-fun apps, which means a practical app might have more of a chance to stand out among the huge number of apps on the App Store.

Other genres that seem popular include weather, book, food and drink, or finance. The book genre seem to overlap a bit with the app idea we described above, but the other genres don't seem too interesting to us:

Weather apps — people generally don't spend too much time in-app, and the chances of making profit from in-app adds are low. Also, getting reliable live weather data may require us to connect our apps to non-free APIs.

Food and drink — examples here include Starbucks, Dunkin' Donuts, McDonald's, etc. So making a popular food and drink app requires actual cooking and a delivery service, which is outside the scope of our company.

Finance apps — these apps involve banking, paying bills, money transfer, etc. Building a finance app requires domain knowledge, and we don't want to hire a finance expert just to build an app.

Now let's analyze the Google Play market a bit.


<a href='#0'> back to index</a>
* * *
## <a id='4'>4. Most Popular Apps by Genre on Google Play</a>

For the Google Play market, we actually have data about the number of installs, so we should be able to get a clearer picture about genre popularity. However, the install numbers don't seem precise enough — we can see that most values are open-ended (100+, 1,000+, 5,000+, etc.):



In [58]:
display_table(free_android_clean, 5) # the Installs columns

1,000,000+ : 15.75 %
100,000+ : 11.54 %
10,000,000+ : 10.57 %
10,000+ : 10.19 %
1,000+ : 8.4 %
100+ : 6.93 %
5,000,000+ : 6.83 %
500,000+ : 5.56 %
50,000+ : 4.77 %
5,000+ : 4.49 %
10+ : 3.54 %
500+ : 3.24 %
50,000,000+ : 2.28 %
100,000,000+ : 2.14 %
50+ : 1.92 %
5+ : 0.79 %
1+ : 0.51 %
500,000,000+ : 0.27 %
1,000,000,000+ : 0.23 %
0+ : 0.05 %
0 : 0.01 %


One problem with this data is that is not precise. For instance, we don't know whether an app with 100,000+ installs has 100,000 installs, 200,000, or 350,000. However, we don't need very precise data for our purposes — we only want to get an idea which app genres attract the most users, and we don't need perfect precision with respect to the number of users.

We're going to leave the numbers as they are, which means that we'll consider that an app with 100,000+ installs has 100,000 installs, and an app with 1,000,000+ installs has 1,000,000 installs, and so on.

To perform computations, however, we'll need to convert each install number to float — this means that we need to remove the commas and the plus characters, otherwise the conversion will fail and raise an error. We'll do this directly in the loop below, where we also compute the average number of installs for each genre (category).

In [59]:
categories_android = freq_table(free_android_clean, 1)
list_order={}

for category in categories_android:
    total = 0
    len_category = 0
    for app in free_android_clean:
        category_app = app[1]
        if category_app == category:            
            n_installs = app[5]
            n_installs = n_installs.replace(',', '')
            n_installs = n_installs.replace('+', '')
            total += float(n_installs)
            len_category += 1
    avg_n_installs = total / len_category
    print(category, ':', avg_n_installs)
    
    list_order[category]=avg_n_installs

ART_AND_DESIGN : 1986335.0877192982
AUTO_AND_VEHICLES : 647317.8170731707
BEAUTY : 513151.88679245283
BOOKS_AND_REFERENCE : 8814199.78835979
BUSINESS : 1712290.1474201474
COMICS : 832613.8888888889
COMMUNICATION : 38590581.08741259
DATING : 854028.8303030303
EDUCATION : 1833495.145631068
ENTERTAINMENT : 11640705.88235294
EVENTS : 253542.22222222222
FINANCE : 1387692.475609756
FOOD_AND_DRINK : 1924897.7363636363
HEALTH_AND_FITNESS : 4188821.9853479853
HOUSE_AND_HOME : 1360598.042253521
LIBRARIES_AND_DEMO : 638503.734939759
LIFESTYLE : 1446158.2238372094
GAME : 15544014.51048951
FAMILY : 3695641.8198090694
MEDICAL : 120550.61980830671
SOCIAL : 23253652.127118643
SHOPPING : 7036877.311557789
PHOTOGRAPHY : 17840110.40229885
SPORTS : 3650602.276666667
TRAVEL_AND_LOCAL : 13984077.710144928
TOOLS : 10830251.970588235
PERSONALIZATION : 5201482.6122448975
PRODUCTIVITY : 16787331.344927534
PARENTING : 542603.6206896552
WEATHER : 5145550.285714285
VIDEO_PLAYERS : 24727872.452830188
NEWS_AND_MAGAZ

In [60]:
marklist = sorted(list_order.items(), key=lambda x:x[1],reverse = True)
sortdict = dict(marklist)
sortdict

{'COMMUNICATION': 38590581.08741259,
 'VIDEO_PLAYERS': 24727872.452830188,
 'SOCIAL': 23253652.127118643,
 'PHOTOGRAPHY': 17840110.40229885,
 'PRODUCTIVITY': 16787331.344927534,
 'GAME': 15544014.51048951,
 'TRAVEL_AND_LOCAL': 13984077.710144928,
 'ENTERTAINMENT': 11640705.88235294,
 'TOOLS': 10830251.970588235,
 'NEWS_AND_MAGAZINES': 9549178.467741935,
 'BOOKS_AND_REFERENCE': 8814199.78835979,
 'SHOPPING': 7036877.311557789,
 'PERSONALIZATION': 5201482.6122448975,
 'WEATHER': 5145550.285714285,
 'HEALTH_AND_FITNESS': 4188821.9853479853,
 'MAPS_AND_NAVIGATION': 4049274.6341463416,
 'FAMILY': 3695641.8198090694,
 'SPORTS': 3650602.276666667,
 'ART_AND_DESIGN': 1986335.0877192982,
 'FOOD_AND_DRINK': 1924897.7363636363,
 'EDUCATION': 1833495.145631068,
 'BUSINESS': 1712290.1474201474,
 'LIFESTYLE': 1446158.2238372094,
 'FINANCE': 1387692.475609756,
 'HOUSE_AND_HOME': 1360598.042253521,
 'DATING': 854028.8303030303,
 'COMICS': 832613.8888888889,
 'AUTO_AND_VEHICLES': 647317.8170731707,
 'L

On average, communication apps have the most installs: 38590581.08741259. 

This number is heavily skewed up by a few apps that have over one billion installs (WhatsApp, Facebook Messenger, Skype, Google Chrome, Gmail, and Hangouts), and a few others with over 100 and 500 million installs:

In [61]:
for app in free_android_clean:
    if app[1] == 'COMMUNICATION' and (app[5] == '1,000,000,000+' or 
                                      app[5] == '500,000,000+' or 
                                      app[5] == '100,000,000+'):
        print(app[0], ':', app[5])

WhatsApp Messenger : 1,000,000,000+
imo beta free calls and text : 100,000,000+
Android Messages : 100,000,000+
Google Duo - High Quality Video Calls : 500,000,000+
Messenger – Text and Video Chat for Free : 1,000,000,000+
imo free video calls and chat : 500,000,000+
Skype - free IM & video calls : 1,000,000,000+
Who : 100,000,000+
GO SMS Pro - Messenger, Free Themes, Emoji : 100,000,000+
LINE: Free Calls & Messages : 500,000,000+
Google Chrome: Fast & Secure : 1,000,000,000+
Firefox Browser fast & private : 100,000,000+
UC Browser - Fast Download Private & Secure : 500,000,000+
Gmail : 1,000,000,000+
Hangouts : 1,000,000,000+
Messenger Lite: Free Calls & Messages : 100,000,000+
Kik : 100,000,000+
KakaoTalk: Free Calls & Text : 100,000,000+
Opera Mini - fast web browser : 100,000,000+
Opera Browser: Fast and Secure : 100,000,000+
Telegram : 100,000,000+
Truecaller: Caller ID, SMS spam blocking & Dialer : 100,000,000+
UC Browser Mini -Tiny Fast Private & Secure : 100,000,000+
Viber Mess

If we removed all the communication apps that have over 100 million installs, the average would be reduced roughly ten times:

In [62]:
under_100_m = []

for app in free_android_clean:
    n_installs = app[5]
    n_installs = n_installs.replace(',', '')
    n_installs = n_installs.replace('+', '')
    if (app[1] == 'COMMUNICATION') and (float(n_installs) < 100000000):
        under_100_m.append(float(n_installs))
        
sum(under_100_m) / len(under_100_m)

3617398.420849421

We see the same pattern for the video players category, which is the runner-up with 24,727,872 installs. The market is dominated by apps like Youtube, Google Play Movies & TV, or MX Player. The pattern is repeated for social apps (where we have giants like Facebook, Instagram, Google+, etc.), photography apps (Google Photos and other popular photo editors), or productivity apps (Microsoft Word, Dropbox, Google Calendar, Evernote, etc.).

Again, the main concern is that these app genres might seem more popular than they really are. Moreover, these niches seem to be dominated by a few giants who are hard to compete against.

The game genre seems pretty popular, but previously we found out this part of the market seems a bit saturated, so we'd like to come up with a different app recommendation if possible.

The books and reference genre looks fairly popular as well, with an average number of installs of 8,767,811. It's interesting to explore this in more depth, since we found this genre has some potential to work well on the App Store, and our aim is to recommend an app genre that shows potential for being profitable on both the App Store and Google Play.

Let's take a look at some of the apps from this genre and their number of installs:

In [63]:
for app in free_android_clean:
    if app[1] == 'BOOKS_AND_REFERENCE':
        print(app[0], ':', app[5])

E-Book Read - Read Book for free : 50,000+
Download free book with green book : 100,000+
Wikipedia : 10,000,000+
Cool Reader : 10,000,000+
Free Panda Radio Music : 100,000+
Book store : 1,000,000+
FBReader: Favorite Book Reader : 10,000,000+
English Grammar Complete Handbook : 500,000+
Free Books - Spirit Fanfiction and Stories : 1,000,000+
Google Play Books : 1,000,000,000+
AlReader -any text book reader : 5,000,000+
Offline English Dictionary : 100,000+
Offline: English to Tagalog Dictionary : 500,000+
FamilySearch Tree : 1,000,000+
Cloud of Books : 1,000,000+
Recipes of Prophetic Medicine for free : 500,000+
ReadEra – free ebook reader : 1,000,000+
Anonymous caller detection : 10,000+
Ebook Reader : 5,000,000+
Litnet - E-books : 100,000+
Read books online : 5,000,000+
English to Urdu Dictionary : 500,000+
eBoox: book reader fb2 epub zip : 1,000,000+
English Persian Dictionary : 500,000+
Flybook : 500,000+
All Maths Formulas : 1,000,000+
Ancestry : 5,000,000+
HTC Help : 10,000,000+
E

The book and reference genre includes a variety of apps: software for processing and reading ebooks, various collections of libraries, dictionaries, tutorials on programming or languages, etc. It seems there's still a small number of extremely popular apps that skew the average:

In [64]:
for app in free_android_clean:
    if app[1] == 'BOOKS_AND_REFERENCE' and (app[5] == '1,000,000,000+'
                                            or app[5] == '500,000,000+'
                                            or app[5] == '100,000,000+'):
        print(app[0], ':', app[5])

Google Play Books : 1,000,000,000+
Bible : 100,000,000+
Amazon Kindle : 100,000,000+
Wattpad 📖 Free Books : 100,000,000+
Audiobooks from Audible : 100,000,000+


However, it looks like there are only a few very popular apps, so this market still shows potential. Let's try to get some app ideas based on the kind of apps that are somewhere in the middle in terms of popularity (between 1,000,000 and 100,000,000 downloads):

In [65]:
for app in free_android_clean:
    if app[1] == 'BOOKS_AND_REFERENCE' and (app[5] == '1,000,000+'
                                            or app[5] == '5,000,000+'
                                            or app[5] == '10,000,000+'
                                            or app[5] == '50,000,000+'):
        print(app[0], ':', app[5])

Wikipedia : 10,000,000+
Cool Reader : 10,000,000+
Book store : 1,000,000+
FBReader: Favorite Book Reader : 10,000,000+
Free Books - Spirit Fanfiction and Stories : 1,000,000+
AlReader -any text book reader : 5,000,000+
FamilySearch Tree : 1,000,000+
Cloud of Books : 1,000,000+
ReadEra – free ebook reader : 1,000,000+
Ebook Reader : 5,000,000+
Read books online : 5,000,000+
eBoox: book reader fb2 epub zip : 1,000,000+
All Maths Formulas : 1,000,000+
Ancestry : 5,000,000+
HTC Help : 10,000,000+
Moon+ Reader : 10,000,000+
English-Myanmar Dictionary : 1,000,000+
Golden Dictionary (EN-AR) : 1,000,000+
All Language Translator Free : 1,000,000+
Aldiko Book Reader : 10,000,000+
Dictionary - WordWeb : 5,000,000+
50000 Free eBooks & Free AudioBooks : 5,000,000+
Al-Quran (Free) : 10,000,000+
Al Quran Indonesia : 10,000,000+
Al'Quran Bahasa Indonesia : 10,000,000+
Al Quran Al karim : 1,000,000+
Al Quran : EAlim - Translations & MP3 Offline : 5,000,000+
Koran Read &MP3 30 Juz Offline : 1,000,000+
H

This niche seems to be dominated by software for processing and reading ebooks, as well as various collections of libraries and dictionaries, so it's probably not a good idea to build similar apps since there'll be some significant competition.

We also notice there are quite a few apps built around the book Quran, which suggests that building an app around a popular book can be profitable. It seems that taking a popular book (perhaps a more recent book) and turning it into an app could be profitable for both the Google Play and the App Store markets.

However, it looks like the market is already full of libraries, so we need to add some special features besides the raw version of the book. This might include daily quotes from the book, an audio version of the book, quizzes on the book, a forum where people can discuss the book, etc.

<a href='#0'> back to index</a>
* * *
## <a id='5'>5. Summary
In this project, we analyzed data about the App Store and Google Play mobile apps with the goal of recommending an app profile that can be profitable for both markets.

We concluded that taking a popular book (perhaps a more recent book) and turning it into an app could be profitable for both the Google Play and the App Store markets. The markets are already full of libraries, so we need to add some special features besides the raw version of the book. This might include daily quotes from the book, an audio version of the book, quizzes on the book, a forum where people can discuss the book, etc.